### PR TITLE
fix: make `String`s in meta backwards compatible

### PR DESCRIPTION
**Pull request by elsirion**
For a long time we allowed plain strings in meta fields, i.e. without some ugly additional JSON quotation, but removed that bug/feature. Turns out that's a backwards incompatible change.

Related to #3648, #3699, #4432 and #4434.

### CODE REVIEW

File: fedimint-core/src/config.rs [hunk 1]
```diff
@@ -198,16 +198,30 @@ impl ClientConfig {
...
+        // In the past we encoded some string fields as "just a string" without quotes,
+        // this code ensures that old meta values still parse since config is hard to
+        // change
+        if res.is_err() && std::any::TypeId::of::<V>() == std::any::TypeId::of::<String>() {
+            let string_ret = Box::new(str_value.clone());
+            let ret: Box<V> = unsafe {
+                // We can transmute a String to V because we know that V==String
+                std::mem::transmute(string_ret)
+            };
+            Ok(Some(*ret))
```
Comment (elsirion) [inline fedimint-core/src/config.rs:221]: I wasn't able to do this with a plain `V`, the compiler complained about unknown size, this seemed like the easiest fix.

Comment (elsirion) [reply fedimint-core/src/config.rs:221]: (@dpc used a similar unsafe trick elsewhere iirc, it's not new)

**Review by douglaz — APPROVED** [review]
Ugly but works

File: fedimint-core/src/config.rs (hunk 1 above)
Comment (bradleystachurski) [reply fedimint-core/src/config.rs:221]: > (@dpc used a similar unsafe trick elsewhere iirc, it's not new)

Yep, there's some insightful discussion related to this pattern in the PR that introduced it https://github.com/fedimint/fedimint/pull/4145

This looks good to me, but I'll defer approval to folks with more context.

File: fedimint-core/src/config.rs [hunk 2]
```diff
@@ -198,16 +198,30 @@ impl ClientConfig {
...
+            .context(format!("Decoding meta field '{key}' failed"));
+
+        // In the past we encoded some string fields as "just a string" without quotes,
+        // this code ensures that old meta values still parse since config is hard to
+        // change
+        if res.is_err() && std::any::TypeId::of::<V>() == std::any::TypeId::of::<String>() {
+            let string_ret = Box::new(str_value.clone());
+            let ret: Box<V> = unsafe {
+                // We can transmute a String to V because we know that V==String
+                std::mem::transmute(string_ret)
```
Comment (dpc) [inline fedimint-core/src/config.rs:219]: Putting bounds here explicitly would make paranoid people feel a bit better about it.

**Review by dpc — APPROVED** [review]

**Pull request merged**